### DISCUSSION TITLE
Markdown tables with captions

### DISCUSSION THREAD

**Discussion by bryanwweber** [executablebooks/meta#201] [Q&A] [answered]
Is it possible to add Markdown tables with captions? I see several pages of documentation for figures that can be referenced, but those seem to require images. There is also a way to glue tabular output from code into Markdown, but I have a table already formatted in Markdown, and I want to add a caption. Thanks!

**Comment by choldgraf** [comment, marked as answer]
Hmmm - can you try wrapping your table in a ` ```{table} ` block like so:

````
```{table} my caption
<table here>
```
````

**Reply by bryanwweber** [reply]
That works! Woo! Thank you 😃  And exactly the command I should have tried >_< I couldn't find it in the docs anywhere though

Is there any way to apply alignment to the table? I tried `:align: center` like a figure, but it didn't change the alignment.

**Reply by chrisjsewell** [reply]
The directive and options is detailed here https://docutils.sourceforge.io/docs/ref/rst/directives.html#table, but note this is **not** a Markdown table, its an rST one which is slightly different.

It is inthe plans to add a markdown one soon though, with similar syntax to the special image syntax:

```
:::{table-md}
a | b
- | -
1 | 2

The caption here
:::
```

**Reply by bryanwweber** [reply]
So, I appreciate the flexibility in the number of approaches you're offering here (and I recognize that each input format needs its own syntax), but the sheer number of options for including content is a little overwhelming, to be honest. For instance, figures can be done with:

````
```{figure} ../images/cool.jpg
---
scale: 50%
align: right
---
Here is my figure caption!
```
````

or

```
:::{figure,myclass} markdown-fig
<img src="../images/fun-fish.png" alt="fishy" class="bg-primary mb-1" width="200px">

This is a caption in **Markdown**!
:::
```

or

````
```{image} ../images/fun-fish.png
:alt: fishy
:class: bg-primary
:width: 200px
:align: center
```
````

or

```
<img src="../images/fun-fish.png" alt="fishy" class="bg-primary" width="200px">
```

or

```
![fishy](../images/fun-fish.png)
```

Which of these should I use? What are the differences between them? I see the last two don't allow a caption or references, but what about the others? I'd personally appreciate a more opinionated set of documentation for what the best practices are or a more opinionated _framework_ overall :smiley:

**Reply by chrisjsewell** [reply]
This is all explained in https://jupyterbook.org/content/figures.html?

**Reply by bryanwweber** [reply]
Yeah, it's all explained there. What I'm trying to express is that I'd prefer that there was _one_ way to do it (in Markdown), or at least, one way that the project says "This is the best practice, do it this way unless you know you need to do it another way"

**Reply by chrisjsewell** [reply]
I'm happy to add a "best practice" if we get a consensus here, but personally I think they all have their use cases and the pros/cons are already outlined in the documentation 😬.

**Reply by bryanwweber** [reply]
So there's three cases where I can't tell what the difference is from reading the docs, other than personal preference about styling:

````
```{image} ../images/fun-fish.png
:alt: fishy
:class: bg-primary
:width: 200px
:align: center
```
````
````
```{figure} ../images/C-3PO_droid.png
---
height: 150px
name: directive-fig
---
Here is my figure caption!
```
````

and

```
:::{figure,myclass} markdown-fig
<img src="../images/fun-fish.png" alt="fishy" class="bg-primary mb-1" width="200px">

This is a caption in **Markdown**!
:::
```

My guess at the difference is that the first syntax is only allowed for images, since images don't have captions. The third may be preferable for Jupyter Notebooks, since the Notebook will display the `<img>` tag while you're working. But that's more-or-less a guess and reading between the lines in the docs, since there's no direct comparison of those three formats.

For code blocks, you can also use the `:tag:` syntax, so I'm not sure if that's also allowed for figures?

Maybe what I'm looking for is a "cheat sheet" or something - if I'm in the middle of writing my book, I just want a quick reference to remind myself "oh, right, that's how I add and reference a figure".

**Reply by bryanwweber** [reply]
By the way, I'm happy to help put something like what I'm suggesting together, if you think it makes sense!

**Reply by chrisjsewell** [reply]
yeh no thanks, I don't want to come across too negative, but I do want to impress that there can be distinct use cases for all of these, and can also depend on the users background; coming from restructedtext or Markdown.
The "markdown-friendly" syntaxes are also quite new functionality (in response to user requests) and so required a bit of time to get feedback before imposing them as "best paractice".

There is already a https://jupyterbook.org/reference/cheatsheet.html, so happy to accept PRs to improve it 👍

**Reply by bryanwweber** [reply]
Ah! That cheat sheet is exactly what I was looking for! Thank you for the link, it helps to read the TOC 🤦
